### PR TITLE
Create exception methods to send error status.

### LINKED ISSUE #12: Send a 400 response to the client if the update organisation request payload is wrong
Write an exception on your own to throw it from the `expect` method. Catch it from the controller and send a 400 response to the user. 
We often don't use built-in exception classes like  `IOExeption`s or `Exception` because they are at the top level in the Exception hierarchy and too generic to determine the exact problem. 
(a good name would be `BadRequestException` within a new package named exception 😛)

### PR DESCRIPTION
**Pull request by pasindur99**
Create exception methods to send error status.

## Purpose
The purpose of this PR is to fix #12

### CODE REVIEW

File: src/main/java/com/OrganizationController.java [hunk 1]
```diff
@@ -1,54 +1,82 @@
...
+            if (id == null) {
+                ServletUtil.respond(organizationService.getAll(), response);
+            } else {
+                Organization organization = organizationService.getOne(Integer.parseInt(request.getParameter("id")));
+                ServletUtil.respond(organization, response);
             }
+        } catch (NotFoundException e) {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            ErrorResource error = new ErrorResource("Bad Request", e.getMessage(), HttpServletResponse.SC_BAD_REQUEST);
+            ServletUtil.respond(error, response);
```
Comment (jayasanka-sack) [inline src/main/java/com/OrganizationController.java:47]: It should be `not found` not `bad request`

Comment (jayasanka-sack) [reply src/main/java/com/OrganizationController.java:47]: applied to other places

File: src/main/java/com/exception/BadRequestException.java [hunk 2]
```diff
@@ -0,0 +1,10 @@
+package com.exception;
+
+public class BadRequestException extends Exception {
+
+    public BadRequestException(String message){
+        super(message);
+    }
+}
+
+
```
Comment (jayasanka-sack) [inline src/main/java/com/exception/BadRequestException.java:10]: remove extra line

File: src/main/java/com/exception/ErrorResource.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+    public int getStatus() {
+        return status;
+    }
+
+    public void setStatus(int status) {
+        this.status = status;
+    }
+}
+
+
```
Comment (jayasanka-sack) [inline src/main/java/com/exception/ErrorResource.java:39]: Remove extra line. Check other files too

File: src/main/java/com/services/OrganizationService.java [hunk 4]
```diff
@@ -56,6 +60,8 @@ public Organization saveOrganization(Organization organization){
                 else {
                     throw new SQLException("Creating organisation failed, no ID obtained.");
                 }
+            } catch (NotFoundException e) {
```
Comment (jayasanka-sack) [inline src/main/java/com/services/OrganizationService.java:63]: Throw without catching

File: src/main/java/com/services/OrganizationService.java [hunk 5]
```diff
@@ -56,6 +60,8 @@ public Organization saveOrganization(Organization organization){
                 else {
                     throw new SQLException("Creating organisation failed, no ID obtained.");
```
Comment (jayasanka-sack) [inline src/main/java/com/services/OrganizationService.java:61]: throw bad request for now

File: src/main/java/com/services/OrganizationService.java [hunk 6]
```diff
@@ -78,10 +84,11 @@ public void deleteOrg (int id){
             throwable.printStackTrace();
```
Comment (jayasanka-sack) [inline src/main/java/com/services/OrganizationService.java:84]: This is the place you should throw the not found I guess.

Comment (jayasanka-sack) [reply src/main/java/com/services/OrganizationService.java:84]: Catch Class not found together with SQL Exception

File: src/main/java/com/services/OrganizationService.java [hunk 7]
```diff
@@ -78,10 +84,11 @@ public void deleteOrg (int id){
...
         } catch (ClassNotFoundException e) {
             e.printStackTrace();
+            throw new NotFoundException("Organization not found for id:" + id);
         }
     }
 
-    public Organization updateOrg (Organization organization) {
+    public Organization updateOrg (Organization organization) throws NotFoundException {
```
Comment (jayasanka-sack) [inline src/main/java/com/services/OrganizationService.java:91]: the method signature says it throws a Not Found exception, but the method never throws.

**Comment by pasindur99** [comment]
Fixed with https://github.com/pasindur99/Organizations-backend/pull/20/commits/bc662658e356d263321b2822a76cdcfa9237164f

File: database/docker-compose.yml [hunk 8]
```diff
@@ -11,4 +11,5 @@ services:
       MYSQL_ROOT_PASSWORD: Admin123
       MYSQL_DATABASE: org
     ports:
-      - 3306:3306
\ No newline at end of file
+      - 3306:3306
+      -
```
Comment (jayasanka-sack) [inline database/docker-compose.yml:15]:
```suggestion

```

File: src/main/java/com/OrganizationController.java [hunk 9]
```diff
@@ -1,54 +1,89 @@
...
+    protected void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        try {
+            Organization organization = ServletUtil.expect(Organization.class, request);
+            ServletUtil.respond(organizationService.saveOrganization(organization), response);
+        } catch (BadRequestException | NotFoundException e) {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            ErrorResource error = new ErrorResource("Bad Request", e.getMessage(), HttpServletResponse.SC_BAD_REQUEST);
+            ServletUtil.respond(error, response);
+        }
```
Comment (jayasanka-sack) [inline src/main/java/com/OrganizationController.java:31]: Catch them separately

File: src/main/java/com/OrganizationController.java [hunk 10]
```diff
@@ -1,54 +1,89 @@
...
+        try {
+            String id = request.getParameter("id");
+            if (id == null) {
+                throw new BadRequestException("Organization doesn't exists for id:"+ id);
+            }else {
+                Organization organization = organizationService.getOne(Integer.parseInt(request.getParameter("id")));
+                organizationService.deleteOrganization(organization);
+            }
+        } catch (BadRequestException e) {
+            e.printStackTrace();
```
Comment (jayasanka-sack) [inline src/main/java/com/OrganizationController.java:64]: Send an error resource to the user

File: src/main/java/com/OrganizationController.java [hunk 11]
```diff
@@ -1,54 +1,89 @@
...
-    protected void doDelete(HttpServletRequest request, HttpServletResponse response) throws ServletException,
-            IOException {
-        int id = Integer.parseInt(request.getParameter("id"));
-        organizationService.deleteOrg(id);
+    protected void doDelete(HttpServletRequest request, HttpServletResponse response) throws IOException {
+
+        try {
+            String id = request.getParameter("id");
+            if (id == null) {
+                throw new BadRequestException("Organization doesn't exists for id:"+ id);
```
Comment (jayasanka-sack) [inline src/main/java/com/OrganizationController.java:58]: This should throw a Bad request.
id == null means user haven't sent an ID. That means it is a bad request.

File: src/main/java/com/OrganizationController.java [hunk 12]
```diff
@@ -1,54 +1,89 @@
...
+            Organization organization = ServletUtil.expect(Organization.class, request);
+            ServletUtil.respond(organizationService.updateOrg(organization), response);
+
+        } catch (BadRequestException e) {
+            response.setStatus(HttpServletResponse.SC_BAD_REQUEST);
+            ErrorResource error = new ErrorResource("Bad Request", e.getMessage(), HttpServletResponse.SC_BAD_REQUEST);
+            ServletUtil.respond(error, response);
+        } catch (NotFoundException e) {
+            response.setStatus(HttpServletResponse.SC_NOT_FOUND);
+            ErrorResource error = new ErrorResource("Not found", e.getMessage(), HttpServletResponse.SC_BAD_REQUEST);
```
Comment (jayasanka-sack) [inline src/main/java/com/OrganizationController.java:85]:
```suggestion
            ErrorResource error = new ErrorResource("Not found", e.getMessage(), HttpServletResponse.SC_NOT_FOUND);
```

File: src/main/java/com/services/OrganizationService.java [hunk 13]
```diff
@@ -92,10 +93,10 @@ public Organization updateOrg (Organization organization) {
             organization = getOne(organization.getId());
         } catch (SQLException throwable) {
             throwable.printStackTrace();
-        } catch (ClassNotFoundException e) {
+        } catch (NotFoundException e) {
             e.printStackTrace();
+            throw new NotFoundException("Organization not found");
```
Comment (jayasanka-sack) [inline src/main/java/com/services/OrganizationService.java:98]: Catch and throw the same exception doesn't makes any sense

File: database/docker-compose.yml (hunk 8 above)
Comment (jayasanka-sack) [reply database/docker-compose.yml:15]: remove `-`

**Comment by pasindur99** [comment]
Fixed with https://github.com/pasindur99/Organizations-backend/pull/20/commits/6619d1aed207904de714d889cd4ef9cd6947f3a4 @jayasanka-sack

**Review by jayasanka-sack — APPROVED** [review]
Well Done! 🥳

**Pull request merged**